osu.Game/IO/OsuStorage.cs

```diff
@@ -9,17 +13,111 @@ namespace osu.Game.IO
             : base(host.Storage, string.Empty)
         {
-            var storageConfig = new StorageConfigManager(host.Storage);
+            this.host = host;
+
+            storageConfig = new StorageConfigManager(host.Storage);
 
             var customStoragePath = storageConfig.Get<string>(StorageConfig.FullPath);
 
             if (!string.IsNullOrEmpty(customStoragePath))
-            {
                 ChangeTargetStorage(host.GetStorage(customStoragePath));
-                Logger.Storage = UnderlyingStorage.GetStorageForDirectory("logs");
+        }
+
+        protected override void ChangeTargetStorage(Storage newStorage)
+        {
+            base.ChangeTargetStorage(newStorage);
+            Logger.Storage = UnderlyingStorage.GetStorageForDirectory("logs");
+        }
+
+        public void Migrate(string newLocation)
+        {
+            var source = new DirectoryInfo(GetFullPath("."));
+            var destination = new DirectoryInfo(newLocation);
+
+            // ensure the new location has no files present, else hard abort
+            if (destination.Exists)
+            {
+                if (destination.GetFiles().Length > 0)
+                    throw new InvalidOperationException("Migration destination already has files present");
+
+                deleteRecursive(destination);
+            }
+
+            copyRecursive(source, destination);
+
+            ChangeTargetStorage(host.GetStorage(newLocation));
+
+            storageConfig.Set(StorageConfig.FullPath, newLocation);
+            storageConfig.Save();
+
+            deleteRecursive(source);
+        }
+
+        private static void deleteRecursive(DirectoryInfo target, bool topLevelExcludes = true)
+        {
+            foreach (System.IO.FileInfo fi in target.GetFiles())
+            {
+                if (IGNORE_FILES.Contains(fi.Name))
+                    continue;
+
+                fi.Delete();
+            }
+
+            foreach (DirectoryInfo dir in target.GetDirectories())
+            {
+                if (IGNORE_DIRECTORIES.Contains(dir.Name))
+                    continue;
+
+                dir.Delete(true);
+            }
+        }
+
+        private static void copyRecursive(DirectoryInfo source, DirectoryInfo destination, bool topLevelExcludes = true)
+        {
+            // based off example code https://docs.microsoft.com/en-us/dotnet/api/system.io.directoryinfo
+            Directory.CreateDirectory(destination.FullName);
+
+            foreach (System.IO.FileInfo fi in source.GetFiles())
+            {
+                if (IGNORE_FILES.Contains(fi.Name))
+                    continue;
+
+                int tries = 5;
+
+                while (tries-- > 0)
+                {
+                    try
+                    {
+                        fi.CopyTo(Path.Combine(destination.FullName, fi.Name), true);
+                        break;
+                    }
+                    catch (Exception)
+                    {
```

Comment:
This should probably have some logging and perhaps completely abort the process if it fails.